### PR TITLE
Reduce memory usage in gKDR

### LINKED ISSUE #242: Optimization in memory usage of gKDR
Hi MOGP developing team, 

I met a RAM exceeding problems when I use gKDR for a 40k dimensional dataset and the it was quite slow to solve a 641d problem. I am raising the issue for a potential optimisation after I compare the gKDR in mogp and its original implementation by matlab. 

I noticed that we compute the matrix R jointly with all samples but this will lead to a problem when the sample size is large. From the original implementation of gKDR, the author uses a loop instead and compute the R incrementally which largely reduce the required memory. 

The matlab code looks like (all the notations are same with mogp):
R = 0;
for i = 1:N
    % Derivative of k(X_i,x) w.r.t x
    Hi = ((repmat(X(i,:),N,1) - X)./SGX/SGX).*repmat(Kx(:,i),1,M);
    %Hi = squeeze(Xij(i,:,:)).*repmat(Kx(:,i),1,M);
    Ri = Hi'*F*Hi;
    R = R+Ri;
end

Best,
Yiming

### PR DESCRIPTION
**Pull request by ots22**
Fixes #242.

The current implementation of gKDR exclusively uses array computations. A drawback of this is that there are some large temporary arrays created (the largest having size `N^2 M^2` where `N` is the number of observations and `M` is the number of features).

This PR replaces one of these array computations with an explicit loop, reducing the space used to `O(M^2 + M N^2)`.

It also replaces two calls to `np.linalg.solve` with calls to `cho_solve` (and a single Cholesky decomposition).

### CODE REVIEW

File: mogp_emulator/DimensionReduction.py [hunk 1]
```diff
@@ -199,19 +200,23 @@ def __init__(self, X, Y, K=None, X_scale = 1.0, Y_scale = 1.0, EPS=1E-8, SGX=Non
         Kx = gram_matrix_sqexp(X, SGX2)
         Ky = gram_matrix_sqexp(Y, SGY2)
 
+        regularized_Kx = Kx + N*EPS*I
+        lower_chol_factor = cholesky(regularized_Kx, lower=True)
```
[the remark this answers is not in the document]
Comment (ots22) [reply mogp_emulator/DimensionReduction.py]: Switched to `cho_factor` as suggested - it makes the calls to `cho_solve` a little cleaner.

File: mogp_emulator/tests/test_DimensionReduction.py [hunk 2]
```diff
@@ -121,3 +121,10 @@ def k_sqexp(x0, x1):
 
     assert(np.allclose(G_sqexp1, G_sqexp_expected))
     assert(np.allclose(G_sqexp2, G_sqexp_expected))
+
+
+def test_DimensionReduction_large():
+    X = np.random.randn(200, 3200)
+    Y = np.arange(200)
+
+    dr = gKDR(X, Y, 2)
```
[the remark this answers is not in the document]
Comment (ots22) [reply mogp_emulator/tests/test_DimensionReduction.py:142]: Yes, just a check that it doesn't crash or error with a moderate problem size (as it does with the original implementation).  Changed to `np.eye` to be on the safe side.

**Comment by ots22** [comment]
@edaub - thanks for the review - I've now finally resolved your comments above.  I've left the target of the PR as main but the version number for you to bump up as discussed.

**Comment by edaub** [comment]
I've decided to stick the three current outstanding bugfixes into a single bugfix release 0.7.2, so changing the base branch to put them all onto a single temporary branch prior to merging. I'll also merge this into `devel`. I don't know if this will fix our issue with `main` and `devel` not knowing their common origin, but I'll try and figure this out separately.

**Pull request merged**